airflow/task/task_runner/standard_task_runner.py

```diff
@@ -186,3 +193,12 @@ def get_process_pid(self) -> int:
         if self.process is None:
             raise RuntimeError("Process is not started yet")
         return self.process.pid
+
+    def _read_task_utilization(self):
+        while True:
+            dag_id = self._task_instance.dag_id
+            task_id = self._task_instance.task_id
+            mem_usage = self.process.memory_percent()
+            cpu_usage = self.process.cpu_percent(interval=1)
```

Comment:
Interesting, `interval=1` is what rate limits this while loop from running too frequently.

I see this in the docs:

>When interval is > 0.0 compares process times to system CPU times elapsed before and after the interval (blocking). When interval is 0.0 or None compares process times to system CPU times elapsed since last call, returning immediately. That means the first time this is called it will return a meaningless 0.0 value which you are supposed to ignore. **In this case is recommended for accuracy that this function be called a second time with at least 0.1 seconds between calls.**

Should we not follow that recommendation? Also is 1 second a bit too tight of a loop? I think that's going to be quite spamy?

Reply:
I am open to change to one second, maybe 5 seconds?

> Should we not follow that recommendation?

If I understand well, the recommendation is when internal is `None`, which is not the case here

Reply:
There is also might be a good idea to grab it inside of `oneshot` context manager
https://psutil.readthedocs.io/en/latest/#psutil.Process.oneshot

In additional it might be also a good idea to catch [NoSuchProcess](https://psutil.readthedocs.io/en/latest/#psutil.NoSuchProcess) and [AccessDenied](https://psutil.readthedocs.io/en/latest/#psutil.AccessDenied) exceptions

I collect example in the single snippet

```python
import multiprocessing as mp
import threading
import time
from string import ascii_letters

import psutil


def _read_task_utilization(pid: int):
    proc = psutil.Process(pid=pid)
    while True:
        try:
            with proc.oneshot():
                mem = proc.memory_percent()
                cpu = proc.cpu_percent()
        except (psutil.NoSuchProcess, psutil.AccessDenied) as ex:
            print(f"Whooops! {type(ex).__name__}: {ex}")
        else:
            print(f"Memory %: {mem}")
            print(f"CPU %: {cpu}")
        time.sleep(1)


def some_expensive_calculation():
    eat_my_memory = ""
    for char in ascii_letters:
        eat_my_memory += char * 1_000_000
        _ = reversed(eat_my_memory)
        time.sleep(0.1)


def main():
    log_reader = threading.Thread(target=_read_task_utilization)
    log_reader.daemon = True
    log_reader.start()

    start = time.monotonic()
    some_expensive_calculation()
    print(f"Total Time: {time.monotonic() - start}")


if __name__ == "__main__":
    mp_context = mp.get_context("fork")
    process = mp_context.Process(target=some_expensive_calculation)
    process.start()

    resource_monitor = threading.Thread(target=_read_task_utilization, args=(process.pid,))
    resource_monitor.daemon = True
    resource_monitor.start()
    process.join()

    time.sleep(3)  # Emulate additional time for a handle after the process finished
```

```console
Memory %: 0.025844573974609375
CPU %: 0.0
Memory %: 0.3806114196777344
CPU %: 1.8
Memory %: 1.0632514953613281
CPU %: 6.3
Memory %: 1.7674446105957031
CPU %: 9.9
Memory %: 1.5843391418457031
CPU %: 11.6
Memory %: 2.2841453552246094
CPU %: 12.7
Whooops! NoSuchProcess: process no longer exists (pid=78454)
Whooops! NoSuchProcess: process no longer exists (pid=78454)
Whooops! NoSuchProcess: process no longer exists (pid=78454)
```

Reply:
And also we might want to handle `AttributeError` since self.process might be changed in MainThread

https://github.com/apache/airflow/blob/5de907504b151eb49e064bb21f5672c142bd595a/airflow/task/task_runner/standard_task_runner.py#L153-L154

https://github.com/apache/airflow/blob/5de907504b151eb49e064bb21f5672c142bd595a/airflow/task/task_runner/standard_task_runner.py#L170-L172

Otherwise daemon could be terminated due to exception and it will be written write it into the logs, which might confuse users

Reply:
Nice idea! I'll do that :)

Reply:
Done! Thanks for the tip @Taragolis. Let me know what you think :)

Reply:
Somehow, `oneshot` makes one unit test fail (not sure which one). I am looking into it

Reply:
Never mind! It was an issue in the unit test